backends/xsmm/ceed-xsmm-tensor.c

```diff
@@ -73,27 +66,112 @@ static int CeedTensorContractApply_Xsmm(CeedTensorContract contract, CeedInt A,
                                         CeedInt B, CeedInt C, CeedInt J,
                                         const CeedScalar *restrict t,
                                         CeedTransposeMode tmode,
-                                        const CeedInt Add,
+                                        const CeedInt add,
                                         const CeedScalar *restrict u,
                                         CeedScalar *restrict v) {
-  CeedInt blksize = 8;
+  int ierr;
+  CeedInt blksize = 8, ind, nelem;
+  CeedTensorContract_Xsmm *impl;
+  ierr = CeedTensorContractGetData(contract, (void *)&impl); CeedChk(ierr);
 
   if (C % blksize)
-    CeedTensorContract_Xsmm_Serial(contract, A, B, C, J, t, tmode, Add, u, v);
+    nelem = 1;
+  else
+    nelem = blksize;
+
+  // Get kernel index
+  if (impl->tensorbasis)
+    ind = CeedGetXsmmInd_Tensor(nelem, add?1:0, tmode==CEED_TRANSPOSE?1:0, B, C,
+                                J, impl->dim);
+  else
+    ind = CeedGetXsmmInd_NonTensor(nelem, add?1:0, impl->P, impl->Q, B, J);
+
+  // Run kernel or fallback to default implementation
+  if (C != 1 && ind < impl->numkernels && ind >= 0 && impl->kernels[ind])
+    for (CeedInt a=0; a<A; a++)
+      impl->kernels[ind](&u[a*B*C], &t[0], &v[a*J*C], NULL, NULL, NULL);
   else
-    CeedTensorContract_Xsmm_Blocked(contract, A, B, C, J, t, tmode, Add, u, v);
+    CeedTensorContract_Xsmm_Default(contract, A, B, C, J, t, tmode, add, u, v);
```

Comment:
We fall back to the default if a valid kernel was not stored or nelem != 1 or 8.

Reply:
Should we error if no valid kernel?

Reply:
Okay, but why does it sometimes return NULL? Are you passing invalid arguments or it just doesn't process some sizes? Since this has performance implications, I'd like for us to understand which code path is taken, not just "randomly" use a slower implementation.